GEODE-1993: postprocess region/key

kjduling opened the pull request:
Add post processing to the GET {region}/{key..key} endpoint
precheckin running

jinmeiliao commented:
A few things:
1. In your RestSecurityPostProcessorTest, in the before method, you can get ahold of the region you created and pre populate the regions with a few keys and values, this way, you don't need to worry about putting in values in your tests.
2. Looks like /{regions} end point also retrieves the values of the entire region, you need to post process the values there as well. 
3. query end points definitely return region values as well.
4. SecurityUtils.getSecurityService will definitely get the security service, but let's do it the spring way, I have a component called RestSecurityService, use that instead.

kjduling commented:
Don't accept this PR yet.  This is just a checkin to preserve work.

Addressed all of the issues listed above.  Added test for adhoc queries.  Still need a test for named queries.  Also need to verify functions won't return any region data without being post-processed.

Have not run precheckin yet, either.

kjduling commented:
precheckin running

kjduling commented:
Precheckin successful except for one flaky test, QueueCommandsDUnitTest.testCreateUpdatesSharedConfig which doesn't appear to be related to this change.

jinmeiliao commented on geode-core/src/main/java/org/apache/geode/security/PostProcessor.java:
```diff
@@ -44,6 +47,9 @@ default void init(Properties securityProps) {}
    */
   Object processRegionValue(Object principal, String regionName, Object key, Object value);
 
+  Collection<Object> processQueryResult(Object principal, Query query, Collection<String> regions,
```
I don't think we want to add this method in the interface. For query,  essentially it's still getting value from region. Customers should not be burdened with another method to implement if they just want to manipulate the data. See how query result get post processed in gfsh queries.

kjduling replied:
removed

jinmeiliao commented on geode-core/src/main/java/org/apache/geode/internal/security/SecurityService.java:
```diff
@@ -87,6 +89,12 @@
   Object postProcess(Object principal, String regionPath, Object key, Object value,
       boolean valueIsSerialized);
 
+  Collection<Object> postProcess(Query query, Collection<String> regionNames,
```
remove this.

kjduling replied:
done

jinmeiliao commented on geode-core/src/main/java/org/apache/geode/internal/security/SecurityService.java:
```diff
@@ -87,6 +89,12 @@
+  Collection<Object> postProcess(Query query, Collection<String> regionNames,
+      Collection<Object> results);
+
+  Collection<Object> postProcess(Object principal, Query query, Collection<String> regionNames,
```
remove this

kjduling replied:
done

jinmeiliao commented on geode-core/src/main/java/org/apache/geode/internal/security/IntegratedSecurityService.java:
```diff
@@ -409,6 +435,34 @@ public Object postProcess(Object principal, String regionPath, Object key, Objec
+  }
+
+  @Override
+  public Collection<Object> postProcess(Object principal, Query query,
```
remove this.

jinmeiliao commented on geode-core/src/main/java/org/apache/geode/internal/security/IntegratedSecurityService.java:
```diff
@@ -409,6 +435,34 @@ public Object postProcess(Object principal, String regionPath, Object key, Objec
+  }
+
+  @Override
+  public Collection<Object> postProcess(Query query, Collection<String> regionNames,
```
remove this.

kjduling replied:
done

jinmeiliao commented on geode-web-api/src/main/java/org/apache/geode/rest/internal/web/controllers/BaseControllerAdvice.java:
```diff
@@ -50,6 +52,11 @@
 
   protected static final String REST_API_VERSION = "/v1";
 
+  public BaseControllerAdvice(final RestSecurityService securityService,
```
do we need this?

kjduling replied:
BaseControllerAdvice extends AbstractBaseController.  Since the RestSecurityService lives in AbstractBaseController, yes.  Otherwise, we define separate member variables in PdxBasedCrudController, CommonCrudController, and QueryAccessContoller.

kjduling commented on geode-assembly/src/test/java/org/apache/geode/rest/internal/web/RestSecurityPostProcessorTest.java:
```diff
@@ -0,0 +1,185 @@
+    HttpResponse response =
+        restClient.doPost("/queries?id=selectCustomer&q=" + URLEncoder.encode(namedQuery, "UTF-8"),
+            "dataReader", "1234567", "");
+    assertEquals(201, getCode(response));
```
I wasn't able to insert it via the API.  When I did, my call to list queries always came up empty and I got a 404 when attempting to run it.

kjduling commented on geode-assembly/src/test/java/org/apache/geode/rest/internal/web/RestSecurityPostProcessorTest.java:
```diff
@@ -0,0 +1,185 @@
+    Region region =
+        serverStarter.cache.createRegionFactory(RegionShortcut.REPLICATE).create(REGION_NAME);
+    region.put("key1",
+        "{\"@type\":\"com.gemstone.gemfire.web.rest.domain.Order\",\"purchaseOrderNo\":1121,\"customerId\":1012,\"description\":\"Order for  XYZ Corp\",\"orderDate\":\"02/10/2014\",\"deliveryDate\":\"02/20/2014\",\"contact\":\"Jelly Bean\",\"email\":\"jelly.bean@example.com\",\"phone\":\"01-2048096\",\"items\":[{\"itemNo\":1,\"description\":\"Product-100\",\"quantity\":12,\"unitPrice\":5,\"totalP ...
```
True.  The tests don't use the object, so I'll remove it for clarity.

kjduling commented on geode-assembly/src/test/java/org/apache/geode/rest/internal/web/RestSecurityPostProcessorTest.java:
```diff
@@ -0,0 +1,185 @@
+    assertEquals(200, getCode(response));
+    assertEquals(MediaType.APPLICATION_JSON_UTF8_VALUE, getContentType(response));
+
+    String customerJson =
```
This breaks from the pattern RestAPIsQueryAndFEJUnitTest follows.  But as you said in an earlier comment about the Orders, it will read cleaner.

kjduling commented:
precheckin successful

jinmeiliao commented:
A few changes requested:

1. use Autowire to auto wire the RestSecurityService.
2. In your RestSecurityPostProcessorTest, considering only using "customers" region for readability. You don't needs all these "/" + REGION_NAME +"/key1" at all.
3. Then you can use a customer domain object specific PostProcessor to change the field values and assert for the value returned.

kjduling commented:
1. I am using Autowire, the annotation is on the base class's constructor.  But as for the injection via the constructor, please review https://spring.io/blog/2015/11/29/how-not-to-hate-spring-in-2016 which references http://olivergierke.de/2013/11/why-field-injection-is-evil/ and http://docs.spring.io/spring-framework/docs/current/spring-framework-reference/htmlsingle/#beans-constructor-injection

Do you think it would be better to only add the RestSecurityService to the controllers that actually need it instead of adding it to the base class?  I think that would complicate the constructors.

2 & 3. Agreed, that'll make the test cleaner.

kjduling commented:
Precheckin successful
